### PR TITLE
Better special characters handling in query escaping

### PR DESCRIPTION
**Pull request by amureki**
Based on PR https://github.com/etianen/django-watson/pull/155

Just updated harmful characters in regexp with MySQL ones, updated tests.

### CODE REVIEW

File: src/watson/backends.py [hunk 1]
```diff
@@ -16,30 +17,28 @@
...
+    """
+    normalizes the query text to a format that can be consumed
     by the backend database
     """
     text = force_text(text)
     text = RE_SPACE.sub(" ", text)  # Standardize spacing.
-    text = RE_NON_WORD.sub("&", text)  # Replace harmful characters with logical "AND"
-    # text may not start or end with "&"
-    text = text.strip('&')
+    text = RE_NON_WORD.sub("", text)  # Remove harmful characters.
```
Comment (etianen) [inline src/watson/backends.py]: One thing - should special characters be replaced by a space, or simply removed? Replacing them with space is equivalent to a logical AND here, whereas removing them simply joins the word. My instinct is that logical AND is safer.

Comment (etianen) [inline src/watson/backends.py]: One thing - should special characters be replaced by a space, or simply removed? Replacing them with space is equivalent to a logical AND here, whereas removing them simply joins the word. My instinct is that logical AND is safer.

**Comment by amureki** [comment]
@etianen yeah, I think you right, better to replace it with spaces.
Made fix, waiting for travis checks.

**Comment by etianen** [comment]
Lovely stuff! Thanks to all involved!

**Comment by JirkaV** [comment]
Thanks all! I'm currently pretty busy at $JOB, so could not make proposed MySQL changes. Thanks @amureki for picking it up!

**Comment by amureki** [comment]
@JirkaV mostly all thanks goes for you, I just wanted to get this as soon as its possible and made small changes. :)
btw, @etianen when do you think, you'll release a new version with all last fixes?

**Comment by JirkaV** [comment]
For the record, the changed version removes some safety belt I put in. If someone searches for a text starting or ending with "&" (which is an allowed char), the search will raise an exception (at least on Postgres). I specifically put in a line of code to prevent that, which was removed by @amureki 

I'd prefer making MySQL-specific version of escape_query() than sacrificing safety, but don't currently have time to do that, so won't complain too much here ;-)

**Comment by etianen** [comment]
Hmm, won't the current code still remove the & and replace it with a space?
Potentially a call to strip() once the character blacklist has been applied
would fix this?

**Comment by etianen** [comment]
Oh, and I can push a new release when I'm back in the office on Monday.

**Comment by JirkaV** [comment]
"Hmm, won't the current code still remove the & and replace it with a space?"

No, the "&" character is one of the "good ones" that are left untouched. So if a user provides it at the beginning or end of their search query, they get an exception. The original pull request accounted for that (and some other edge cases).

**Comment by amureki** [comment]
@JirkaV so, can you, please, tell once again what other edge cases were there?
So if you don't have time, I can fix it back and also that one with "&" symbol.

**Comment by JirkaV** [comment]
Absolutely. Don't get me wrong, I'm happy that you converted my pull request to a working version. I only had a chance to fine-tune it for PostgreSQL. The tests were not broken on MySQL so I assumed it was good enough.

Anyway, if you do something like this in PostgreSQL ```select to_tsquery('aaa&')```, you get error "no operand in tsquery". So the query that's being sent to the DB may not start or end with "&". There was a special line in my original code handling that - basically a simple strip() call.

HTH

   Jirka

**Pull request merged**